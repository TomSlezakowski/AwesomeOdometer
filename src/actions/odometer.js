export const RANDOMIZE_NUMBER_SET = 'RANDOMIZE_NUMBER_SET'
export function randomizeNumberSet(randomNumber) {
  return {
    type: RANDOMIZE_NUMBER_SET,
    randomNumber
  }
}