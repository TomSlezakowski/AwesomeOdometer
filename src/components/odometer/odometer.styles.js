import styled, { keyframes } from 'react-emotion'

const slide = ( viewport ) => keyframes({
  '0%': {
    transform: 'translateY(0)'
  },
  '100%': {
    transform: `translateY(-100%) translateY(${viewport}em)`
  },
})

export const OdometerStyled = styled('div')(
  {
    display: 'inline-block',
    overflow: 'hidden',
  },
  ({ theme }) => {
    switch (theme) {
      case 'light':
        return {
          background: '#F4F4F4',
          boxShadow: 'inset 0 2px 8px -2px #000',
          color: 'rgba(0, 0, 0, 0.9)',
          borderRadius: '2px',
        }
      case 'dark':
        return {
          background: '#222',
          boxShadow: 'inset 0 2px 8px -2px #000',
          color: 'rgba(255, 255, 255, 0.9)',
          borderRadius: '2px',
        }
      default:
        return {}
    }
  }
)

export const DigitStyled = styled('div')(
  {
    display: 'inline-block',
    position: 'relative',
    boxSizing: 'unset',
  },
  ({ viewport }) => ({
    height: `${viewport}em`,
  }),
  ({ isDecimal }) => ({
    width: isDecimal ? '0.3em' : '0.6em'
  }),
  ({ theme }) => {
    switch (theme) {
      case 'light':
        return {
          borderRight: '1px solid rgba(255, 255, 255, 0.5)',
          borderLeft: '1px solid rgba(0, 0, 0, 0.05)',
          padding: '0 0.2rem',
          '::first-child': {
            borderLeft: 'none'
          },
        }
      case 'dark':
        return {
          borderRight: '1px solid rgba(0, 0, 0, 0.5)',
          borderLeft: '1px solid rgba(255, 255, 255, 0.05)',
          padding: '0 0.2rem',
          '::first-child': {
            borderLeft: 'none'
          },
        }
      default:
        return {}
    }
  },
  ({ customStyling }) => ({
    borderLeft: customStyling.borderLeft,
    borderRight: customStyling.borderRight,
    padding: customStyling.padding
  }),
)

export const DigitContainerStyled = styled('div')(
  {
    overflow: 'hidden',
    animationFillMode: 'forwards',
    display: 'flex',
    position: 'relative',
    animationDelay: '1000ms',
  },
  ({ viewport }) => ({
    lineHeight: viewport,
  }),
  ({ startAnim, viewport }) => ({
    animationName: startAnim ? `${slide(viewport)}` : 'none',
  }),
  ({ turnCount, animation, easing })  => {
    switch (animation) {
      case 'turn':
        return {
          animationTimingFunction: easing
        }
      case 'step':
        return {
          animationTimingFunction: `steps(${turnCount}, end)`
        }
      default:
        return {
          animationTimingFunction: easing
        }
    }
  },
  ({ duration }) => ({
    animationDuration: `${duration}ms`,
  })
)

