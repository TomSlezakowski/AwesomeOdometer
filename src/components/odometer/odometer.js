import * as React from 'react'
import { OdometerStyled, DigitStyled, DigitContainerStyled } from './odometer.styles'

/* ****************************************** */
/*

  Usage:
    <Odometer start={12.50} end={26.51} duration={700} animation={'turn'} />

  Options:
    start: start value
    end: end value
    prefix: string value shown before value
    duration: animation duration
      // 1000 [Default]
    animation: animation types
      // turn [Default] - rotates digits
      // step - changes digits
    maxTurnCount: maximum number of turns for each digit
      // '10' [Default] (+ index*2, so each digit moves out of sync)
      // '0' - max turn count - warning, can be in the thousands if start/end value is more than 4 digits eg. 10.00
      // 'customvalue'  (+ index*2, so each digit moves out of sync)
    startAnim: set to true to start
      // false [Default]
      // true
    viewport: gap in which value can be seen rotating
      // 1.7 [Default]
    easing: custom animation timing
      // ease-out [Default]
      // ease-in
      // linear
      // custom eg: cubic-bezier(0,0,1,-1.49)
    theme: color theme
      // no theme [default]
      // dark
      // light
      // custom {background, color, border}
    custom: customize parts of the theme
      // borderLeft
      // borderRight
      // padding

*/
/* ****************************************** */

const NumString = ({ start, end, turnCount, maxTurns }) => {
  // Generates the number string for rotating number
  // Using maxTurnCount to limit length of string generted
  // When the string is too long the number rotates so fast that it is perceived to rotate backwards

  let numberStr = ''
  let countRunner = start
  const turns = maxTurns ? maxTurns : turnCount
  const tickAddition = turns !== 0 ? Math.floor(turnCount / turns) : 0

  // returns start if there are no turns
  if (turns !== 0) {
    numberStr = start + ' '
    for (let i = 0; i < turns - 1; i++)  {
      countRunner = countRunner += tickAddition
      numberStr += (Math.floor(countRunner) % 10) + ' '
    }
    numberStr += end + ''
  } else {
    numberStr = start + ''
  }

  return numberStr
}

const GetTurns  = ({ startDigitsMatched, difference }) =>  {
  // Calculates how many turns each digit needs to make to move from start to end number

  const turnsArr = []
  const newstartDigitsMatched = [...startDigitsMatched]
  const startDigitMatchedCount = startDigitsMatched.length
    startDigitsMatched.map((digit, index) => {
    // remove first item from array
    newstartDigitsMatched.shift()
    if (newstartDigitsMatched.length  === 0) {
      newstartDigitsMatched.push('0')
    }
    // merge array into string then convert to number
    let startJoined = parseInt(newstartDigitsMatched.join(''), 10)

    // What is the current turn rotation of the current digit  eg. 1500 the turn rotation of the first digit is 0.5, where 0 is 0% and 1 is 100%
    // currentTurnRotation.push(startJoined / Math.pow(10, startDigitMatchedCount - 1 - index))
    // Turn rotations needed to move to next digit
    // differenceArr.push(difference / Math.pow(10, startDigitMatchedCount - 1 - index))
    turnsArr.push(Math.floor((startJoined / Math.pow(10, startDigitMatchedCount - 1 - index)) + (difference / Math.pow(10, startDigitMatchedCount - 1 - index))))
    return null
  })
  return turnsArr
}

export const Odometer = ({ start, end, duration = 1000, animation = 'turn', maxTurnCount = 10, startAnim = true, prefix = '', viewport = 1.7, easing = 'ease-out', theme = 'none', customStyling =  {} }) => {

  // Math.round prevents weird numbers from appearing
  // eg multiplying 17.51 by 100 produces 1751.0000000000002
  const direction = end > start ? 'forward' : 'backward'
  // if direction is 'backwards' reverse start and end values
  const startCopy = direction === 'forward' ? Math.round(start * 100) : Math.round(end * 100)
  const endCopy = direction === 'forward' ? Math.round(end * 100) : Math.round(start * 100)
  const startDigits = startCopy.toString().split('')
  const endDigits = endCopy.toString().split('')
  const difference = Math.abs(endCopy - startCopy)
  const startCount = startDigits.length
  const endCount = endDigits.length
  const countDifference = endCount - startCount
  const digitCount = Math.max(startCount, endCount)
  let startDigitsMatched = [...startDigits]
  let endDigitsMatched = [...endDigits]

  // Ensure both start and end have the same number of digits
  // If start/end has fewer digits, add 0 in front
  if (countDifference > 0) {
    startDigitsMatched = Array(Math.abs(countDifference)).fill(0).concat(startDigits)
  }
  if (countDifference < 0) {
    endDigitsMatched = Array(Math.abs(countDifference)).fill(0).concat(endDigits)
  }

  const turnsArr = GetTurns({ startDigitsMatched, difference })
  const rows = []

  // Prefix
  if (prefix !== '') {
    rows.push(
      <DigitStyled key={'prefix'} viewport={viewport} theme={theme} customStyling={customStyling}>
        <DigitContainerStyled count={digitCount - 1} turnCount={0} direction={direction} startAnim={startAnim} viewport={viewport} easing={easing} theme={theme}>{prefix}</DigitContainerStyled>
      </DigitStyled>
    )
  }

  for (let index = 0; index < digitCount; index++) {
    // Capped at 1000 turns
    const turnCount = turnsArr[index] > 1000 ? 1000 : turnsArr[index]
    const maxTurns = Math.min((maxTurnCount === 0 ? turnCount : maxTurnCount + index * 2), turnCount)

    rows.push(
      <DigitStyled key={index} viewport={viewport} theme={theme} customStyling={customStyling}>
        <DigitContainerStyled count={digitCount - 1} turnCount={maxTurns} direction={direction} animation={animation} duration={duration} startAnim={startAnim} viewport={viewport} easing={easing} theme={theme}>
          {NumString({ start: parseInt(startDigitsMatched[index], 10), end: parseInt(endDigitsMatched[index], 10), turnCount: turnsArr[index], maxTurns })}
        </DigitContainerStyled>
      </DigitStyled>
    )

    // Decimal point added 2 digits from the end
    if (index === digitCount - 3) {
      rows.push(
        <DigitStyled isDecimal={true} key='decimal' viewport={viewport} theme={theme} customStyling={customStyling}>
          <DigitContainerStyled count={digitCount - 1} turnCount={0} direction={direction} startAnim={startAnim} viewport={viewport} easing={easing} theme={theme}>.</DigitContainerStyled>
        </DigitStyled>
      )
    }
  }

  return (
    <OdometerStyled theme={theme}>{rows}</OdometerStyled>
  )
}
