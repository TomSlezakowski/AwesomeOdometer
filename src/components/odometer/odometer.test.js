import * as React from 'react'
import { mount, ReactWrapper } from 'enzyme'
import 'jest-enzyme'
import { Odometer } from './odometer.js'
import { DigitContainerStyled } from './odometer.styles'

describe('Odometer component tests', () => {
  describe('When mounted', () => {
    let component

    beforeEach(() => {
      component = mount(<Odometer start={4.55} end={5.10} duration={700} action={'turn'} startAnim={true} prefix={'$'} />)
    })

    afterEach(() => {
      component.unmount()
      component = null
    })

    it('Renders correctly', () => {
      expect(component).toMatchSnapshot()
    })

    it('Should contain correct value', () => {
      expect(component.find(DigitContainerStyled).at(1)).toHaveText('4 5')
      expect(component.find(DigitContainerStyled).at(2)).toHaveText('.')
      expect(component.find(DigitContainerStyled).at(3)).toHaveText('5 6 7 8 9 0 1')
      expect(component.find(DigitContainerStyled).at(4)).toHaveText('5 0 6 1 7 2 8 3 9 4 0')
    })
  })
})
