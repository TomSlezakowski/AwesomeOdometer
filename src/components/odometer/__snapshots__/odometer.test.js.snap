// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Odometer component tests When mounted Renders correctly 1`] = `
<Component
  action="turn"
  duration={700}
  end={5.1}
  prefix="$"
  start={4.55}
  startAnim={true}
>
  <Styled(div)>
    <div
      className="css-i6bazn"
    >
      <Styled(div)
        key="prefix"
      >
        <div
          className="css-1d4o148"
        >
          <Styled(div)
            count={2}
            startAnim={true}
            turnCount={0}
          >
            <div
              className="css-jyp0ze"
            >
              $
            </div>
          </Styled(div)>
        </div>
      </Styled(div)>
      <Styled(div)
        key="0"
      >
        <div
          className="css-1d4o148"
        >
          <Styled(div)
            action="turn"
            count={2}
            duration={700}
            startAnim={true}
            turnCount={1}
          >
            <div
              action="turn"
              className="css-d23gjp"
            >
              4 5
            </div>
          </Styled(div)>
        </div>
      </Styled(div)>
      <Styled(div)
        isDecimal={true}
        key="decimal"
      >
        <div
          className="css-19hk81p"
        >
          <Styled(div)
            count={2}
            startAnim={true}
            turnCount={0}
          >
            <div
              className="css-jyp0ze"
            >
              .
            </div>
          </Styled(div)>
        </div>
      </Styled(div)>
      <Styled(div)
        key="1"
      >
        <div
          className="css-1d4o148"
        >
          <Styled(div)
            action="turn"
            count={2}
            duration={700}
            startAnim={true}
            turnCount={6}
          >
            <div
              action="turn"
              className="css-d23gjp"
            >
              5 6 7 8 9 0 1
            </div>
          </Styled(div)>
        </div>
      </Styled(div)>
      <Styled(div)
        key="2"
      >
        <div
          className="css-1d4o148"
        >
          <Styled(div)
            action="turn"
            count={2}
            duration={700}
            startAnim={true}
            turnCount={55}
          >
            <div
              action="turn"
              className="css-d23gjp"
            >
              5 0 6 1 7 2 8 3 9 4 0
            </div>
          </Styled(div)>
        </div>
      </Styled(div)>
    </div>
  </Styled(div)>
</Component>
`;
